### PR TITLE
nix: Add script for auto-regenerating and fix the build

### PR DESCRIPTION
**Pull request by rvl**
- This will allow Buildkite to regenerate nix files and then push them to your PR branch.

   For the push to work, we need to add a ssh deployment key with write access on this GitHub repo.

- There is also a fix for the nix build since `cardano-wallet-launcher` was removed.

- And there are also some tidy ups I found while writing the [Building](https://github.com/input-output-hk/cardano-wallet/wiki/Building) page on the wiki.

### CODE REVIEW

File: lib/core/cardano-wallet-core.cabal [hunk 1]
```diff
@@ -51,6 +51,7 @@ library
     , iohk-monitoring
     , memory
     , monad-logger
+    , mtl
```
Comment (KtorZ) [inline lib/core/cardano-wallet-core.cabal]: Is this an oversight :thinking:  ?

File: nix/check-nix-tools.nix [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  else
+    echo "Committing changes..."
+    commit_message="Regenerate nix"
+
+    # If on a PR branch, search for a previous regen commit to fix up.
+    commit_fixup=""
+    if [ -n "''${BUILDKITE_PULL_REQUEST_BASE_BRANCH:-}" ]; then
+      previous=$(git log --pretty=oneline --no-decorate origin/$BUILDKITE_PULL_REQUEST_BASE_BRANCH..HEAD | awk "/$commit_message/ { print \$1; }")
+      if [ -n $previous ]; then
+        commit_fixup="--fixup $previous"
```
Comment (KtorZ) [inline nix/check-nix-tools.nix]: :+1:

File: nix/check-nix-tools.nix [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+      fi
+    fi
+
+    # Create the commit
+    export GIT_COMMITTER_NAME="IOHK"
+    export GIT_COMMITTER_EMAIL="devops+nix-tools@iohk.io"
+    export GIT_AUTHOR_NAME="$GIT_COMMITTER_NAME"
+    export GIT_AUTHOR_EMAIL="$GIT_COMMITTER_EMAIL"
+    git commit --no-gpg-sign $commit_fixup --message "$commit_message"
```
Comment (KtorZ) [inline nix/check-nix-tools.nix]: We do at the moment requires commit to be signed. Perhaps it possible to fill an exception for `devops+nix-tools@iohk.io`.

**Review by KtorZ — APPROVED** [review]

File: lib/core/cardano-wallet-core.cabal (hunk 1 above)
Comment (KtorZ) [reply lib/core/cardano-wallet-core.cabal]: Ah -->  `Temporary commit -- tests the buildkite job`

File: nix/check-nix-tools.nix (hunk 3 above)
Comment (rvl) [reply nix/check-nix-tools.nix]: Commit signing is possible. It means deploying a GPG key to the buildkite agents and then adding the key to a GitHub user such as @iohk-devops.

**Comment by rvl** [comment]
I have rebased on latest master and removed the temporary commit that causes regeneration.

How should we proceed with this?

The commit that Buildkite makes looks like this: e671871028905014842cdc4d6e1b47303b20be6b

They commits aren't GPG signed. For signed commits, there is a little bit of devops effort to automate GPG key deployment on the build agents. Then (as I recall) the git invocation might need a couple of environment variables set to work around pin entry and gpg-agent things.

**Review by KtorZ — APPROVED** [review]
LGTM. I'd have to disable the requirement on signed commit though :/

**Comment by KtorZ** [comment]
As discussed:

Keeping this open until we find a way to GPG-signed commits.

**Comment by KtorZ** [comment]
Merging as-is. It's unclear when we'll get GPG signed commit with that (not sure if there's a ticket somewhere?) but we may disable that branch protection since that's also a problem for bors. 
At least, every developer is now signing commits.

**Pull request merged**